src/main/java/com/googleintern/wfm/ruleengine/action/RuleValidation.java

```diff
@@ -0,0 +1,286 @@
+    return roleIdBuilder.toString();
+  }
+
+  private static String convertSkillIdsToString(List<Long> skillIds, List<Long> roleSkillIds) {
+    StringBuilder skillIdBuilder = new StringBuilder();
+    skillIdBuilder.append(Separator.SQUARE_BRACKET_LEFT.symbol);
+    for (Long skillId : skillIds) {
+      skillIdBuilder.append(skillIdBuilder.length() == 1 ? "" : Separator.COMMA.symbol);
+      skillIdBuilder.append(skillId);
+    }
+    for (Long roleSkillId : roleSkillIds) {
+      skillIdBuilder.append(skillIdBuilder.length() == 1 ? "" : Separator.COMMA.symbol);
+      skillIdBuilder.append(roleSkillId);
+    }
+    skillIdBuilder.append(Separator.SQUARE_BRACKET_RIGHT.symbol);
+    return skillIdBuilder.toString();
+  }
+
+  private static String convertPoolAssignmentsToString(Set<PoolAssignmentModel> poolAssignments) {
+    StringBuilder poolAssignmentsStringBuilder = new StringBuilder();
+    poolAssignmentsStringBuilder.append(Separator.SQUARE_BRACKET_LEFT.symbol);
+    for (PoolAssignmentModel poolAssignment : poolAssignments) {
+      poolAssignmentsStringBuilder.append(
+          poolAssignmentsStringBuilder.length() == 1 ? "" : Separator.COMMA.symbol);
+      poolAssignmentsStringBuilder.append(
+          CASE_POOL_ID_PREFIX
+              + poolAssignment.casePoolId()
+              + PERMISSION_SET_ID_PREFIX
+              + poolAssignment.permissionSetId()
+              + Separator.DOUBLE_QUOTATION_MARK.symbol
+              + Separator.CURLY_BRACKET_RIGHT.symbol);
+    }
+    poolAssignmentsStringBuilder.append(Separator.SQUARE_BRACKET_RIGHT.symbol);
+    return poolAssignmentsStringBuilder.toString();
+  }
+
+  private static ImmutableSetMultimap<UserPoolAssignmentModel, PoolAssignmentModel>
+      assignPermissionsBasedOnGeneratedRules(
+          List<UserPoolAssignmentModel> expectedUserPoolAssignments) {
+    ImmutableSetMultimap.Builder<UserPoolAssignmentModel, PoolAssignmentModel>
+        filtersByUserPoolAssignmentBuilder = ImmutableSetMultimap.builder();
+    for (UserPoolAssignmentModel user : expectedUserPoolAssignments) {
+      ImmutableSet<FilterModel> userFilters =
+          ImmutableSet.copyOf(
+              CasePoolIdAndPermissionIdGroupingUtil.convertSkillIdRoleIdToFilter(user));
+      for (RuleModel rule : rules) {
+        ImmutableSet.Builder<PoolAssignmentModel> assignedPoolAssignmentsBuilder =
+            ImmutableSet.builder();
+        if (decideToAssignPermissions(user.workforceId(), user.workgroupId(), userFilters, rule)) {
+          for (Long permissionSetId : rule.permissionSetIds()) {
+            assignedPoolAssignmentsBuilder.add(
+                PoolAssignmentModel.builder()
+                    .setCasePoolId(rule.casePoolId())
+                    .setPermissionSetId(permissionSetId)
+                    .build());
+          }
+        }
+        filtersByUserPoolAssignmentBuilder.putAll(user, assignedPoolAssignmentsBuilder.build());
+      }
+    }
```

Comment:
Avoid using loops on top of other loops for too many layers. 
Either simplified them using lambda function or take the logic out to another helper method.